File: src/examples/decode.js
let {
    Codec,
    TypeFactory,
    Utils,
    Types,
    WireTypes
} = require('../index')

let codec = new Codec();

let A = TypeFactory.create('A', [ {
   name: "a",
    type: Types.String
},
{
    name: "b",
    type: Types.Int8
},
{
    name: "c",
    type: Types.Struct
}
])

let SubA = TypeFactory.create('SubA', [{
    name: "a",
    type: Types.String
},
{
    name: "b",
    type: Types.Int8
},
{
    name: "c",
    type: Types.Struct
}
])

let SubB = TypeFactory.create('SubB', [{
    name: "a",
    type: Types.String
}
])

const toHex = (buffer) => {
    let s = ''
    buffer.forEach((b) => {
        b = b.toString(16)
        if (b.length == 1) {
            b = '0' + b
        }
        s += b
    })
    return s
  }
  
  const fromHex = (str) => {
    let buffer = []
    for (let i = 0; i < str.length; i += 2) {
      let hex = str.slice(i, i+2)
      buffer.push(parseInt(hex, 16))
    }
    return buffer
  }

codec.registerConcrete(new A(), "A", {})
codec.registerConcrete(new SubA, "SubA",{})
codec.registerConcrete(new SubB, "SubB",{})
let subB = new SubB("Do Ngoc Tan")
let subA =  new SubA("Hello World",5, subB)
let a = new A("Tan",45,subA)


let binary = codec.marshalBinary(a)
console.log("binary.length=",binary.length)
//let decodedData = new A()
let decodedData = new A()
//console.log("A=",new A())
//console.log("SubB=",new SubB())
//console.log("subA=",new SubA())
codec.unMarshalBinary(binary, decodedData)
console.log(decodedData.JsObject())

File: src/examples/decode.go
package main

import (
	"encoding/hex"
	"fmt"

	amino "github.com/tendermint/go-amino"
)

func main() {

	type SubB struct {
		C string
	}

	type SubA struct {
		Str string
		B   int8
		C   SubB
	}

	type A struct {
		Str1 string
		Num1 int8
		Sub  SubA
	}

	var cdc = amino.NewCodec()

	cdc.RegisterConcrete(A{}, "A", nil)
	cdc.RegisterConcrete(SubA{}, "SubA", nil)
	cdc.RegisterConcrete(SubB{}, "SubB", nil)
	/*
		s := SimpleStruct{Sub: SubStruct{A: 5}}
		b, err := cdc.MarshalBinary(s)
		if err == nil {
			fmt.Println("b=", b)
			//fmt.Println(string(s.D.E))
		} else {
			fmt.Println(err)
		}*/
	bz, _ := hex.DecodeString("34d08264d30a0354616e105a1b7a49d00b0a0b48656c6c6f20576f726c64100a1bdb8a5de30a0b446f204e676f632054616e040404")
	var s2 A
	cdc.UnmarshalBinary(bz, &s2)
	fmt.Println(s2)

}
